include/s3select_json_parser.h

```diff
@@ -0,0 +1,383 @@
+
+    JsonParserHandler() : init_buffer_stream(false)
+    {} 
+
+    std::string get_key_path()
+    {//for debug
+	  std::string res;
+	  for(const auto & i: key_path)
+	  {
+	    res.append(i);
+	    res.append(std::string("/"));
+	  }
+	  return res;
+    }
+
+    void dec_key_path()
+    {
+      if(json_element_state.back() != ARRAY_STATE)
+	    {
+	      if(key_path.size() != 0) {
+	      key_path.pop_back();
+        }
+	    }
+      if (key_path.size() < from_clause.size()) {
+        flag = false;
+        } 
+    }
+
+    void push_new_key_value(Valuesax& v)
+    {  
+      if (flag) {
+        for (size_t i = 1; i < query_matrix.size(); i++) { // 0th index -> from-clause. Ignore that.
+          if(std::equal(key_path.begin() + from_clause.size(), key_path.end(), query_matrix[i].begin())) {
+            fp(make_pair(get_key_path(),v));
+          }
+        }
+      }
+      dec_key_path();
+    }
+
+    bool Null() {
+      // at this point should verify against from-clause/where-clause/project. if match then push to scratch-area
+      push_new_key_value(value.Parse(nullptr));
+      return true; }
+
+    bool Bool(bool b) {
+      push_new_key_value(value.Parse(b));
+      return true; }
+
+    bool Int(int i) { 
+      push_new_key_value(value.Parse(i));
+      return true; }
+
+    bool Uint(unsigned u) {
+      push_new_key_value(value.Parse(u));
+      return true; }
+
+    bool Int64(int64_t i) { 
+      push_new_key_value(value.Parse(i));
+      return true; }
+
+    bool Uint64(uint64_t u) { 
+      push_new_key_value(value.Parse(u));
+      return true; }
+
+    bool Double(double d) { 
+      push_new_key_value(value.Parse(d));
+      return true; }
+
+    bool String(const char* str, rapidjson::SizeType length, bool copy) {
+      //TODO use copy
+      push_new_key_value(value.Parse(str));
+      return true;
+    }
+
+    bool Key(const char* str, rapidjson::SizeType length, bool copy) {
+      key_path.push_back(std::string(str));
+      
+      if (key_path == from_clause) {
+        flag = true;
+      }
+      return true;
+    }
+
+    bool StartObject() {
+      json_element_state.push_back(OBJECT_STATE);
+      return true; 
+    }
+  
+    bool EndObject(rapidjson::SizeType memberCount) {
+      json_element_state.pop_back();
+      dec_key_path();
+
+      return true; 
+    }
+ 
+    bool StartArray() {
+      json_element_state.push_back(ARRAY_STATE);
+      return true;
+    }
+
+    bool EndArray(rapidjson::SizeType elementCount) { 
+      json_element_state.pop_back();
+      dec_key_path();
+      return true;
+    }
+
+    void set_function(std::function <int(std::pair < std::string, Valuesax>)> f)
+    {
+      fp = f;
+    }
+
+    int process_json_buffer(char* json_buffer,size_t json_buffer_sz, std::function <int(std::pair < std::string, Valuesax>)> f, bool end_of_stream=false)
+    {//user keeps calling with buffers, the method is not aware of the object size.
+
+      set_function(f);
```

Comment:
there is no need to call to `set_function` while `f` reside on `process_json_buffer` interface.
`process_json_buffer` is called per new incoming chunk